src/Hopper/Internal/LoweredCore/EvalClosureConvertedANF.hs

```diff
@@ -486,47 +513,102 @@ enterTotalMathPrimopSimple
   -> (GmpMathOpId, V.Vector Ref)
   -> forall c. EvalCC c s (V.Vector Ref)
 enterTotalMathPrimopSimple controlstack (opId, refs) = do
-  traceShowM refs
   heapHandle <- _extractHeapCAH <$> getHSCM
-  traceShowM heapHandle
-  traceShowM $ V.length refs
-  args <- extendErrorTrans $ mapM heapLookup refs
+  args <- mapM heapLookup refs
   checkedOp <- return $ do
     areLiterals <- mapM argAsLiteral (V.toList args)
     hoistTotalMathLiteralOp opId areLiterals
   case checkedOp of
     Left str ->
-      let errMsg = "Received a non-literal to a total math primop :\n" ++ str
+      let errMsg = "Received a non-literal to a total math primop:\n" ++ str
           err step = PanicMessageConstructor(controlstack, 0, InterpreterStepCC step, errMsg)
       in throwEvalError err
     Right litOp -> do
-      _ <- extendErrorTrans $ checkedCounterCost (fromIntegral $ gmpMathCost litOp)
+      _ <- checkedCounterCost (fromIntegral $ gmpMathCost litOp)
       lits <- return $ evalTotalMathPrimopCode litOp
       mapM allocLit $ V.fromList lits
 
   where argAsLiteral :: ValueRepCC Ref -> Either String Literal
         argAsLiteral (ValueLitCC lit) = Right lit
         argAsLiteral notLit = Left $ "Not a literal: `" ++ show notLit ++ "`"
 
+-- | Enter the top of the stack, adding some refs to scope.
 enterControlStackCC
   :: SymbolRegistryCC
   -> ControlStackCC
   -> V.Vector Ref
   -> forall c. EvalCC c s (V.Vector Ref)
 enterControlStackCC _ ControlStackEmptyCC refs = return refs
 enterControlStackCC registry stack@(UpdateHeapRefCC ref newStack) refs = do
-  val <- extendErrorTrans $ heapLookup ref
-  if val == BlackHoleCC
-    then do
-      extendErrorTrans $ unsafeHeapUpdate ref (IndirectionCC refs)
-      enterControlStackCC registry newStack refs
-    else do
-      errMsg <- return $  unwords
-            [ "UpdateHeapRefCC expected a black hole instead of"
-            , "`" ++ show val ++ "`"
-            ]
-      throwEvalError
-        (\step ->  PanicMessageConstructor(stack, 0, InterpreterStepCC step, errMsg))
+  val <- heapLookup ref
+  unsafeHeapUpdate ref (IndirectionCC refs)
+  enterControlStackCC registry newStack refs
+  -- if val == BlackHoleCC then do
+  --   else do
+  --     errMsg <- return $  unwords
+  --           [ "UpdateHeapRefCC expected a black hole instead of"
+  --           , "`" ++ show val ++ "`"
+  --           ]
+  --     throwEvalError
+  --       (\step ->  PanicMessageConstructor(stack, 0, InterpreterStepCC step, errMsg))
 enterControlStackCC registry (LetBinderCC _binders () newEnv body newStack) refs = do
   envStack <- return (EnvConsCC refs newEnv)
   evalCCAnf registry envStack newStack body
+
+
+-- Some quick typedefs to make the `lookupStaticValue` reader stuff easier.
+type SymbolTable = Map.Map GlobalSymbol (ValueRepCC GlobalSymbol)
+
+type EvalCCT s
+  = HeapStepCounterM (ValueRepCC Ref)
+                     (STE SomeHopperException s)
+
+type Inflate a = forall c s. ReaderT SymbolTable (EvalCCT s) a
```

Comment:
theres probably some bigger picture things to think about here, but that `forall c` is unused :)